crates/vm/levm/src/call_frame.rs

```diff
@@ -28,29 +28,33 @@ pub struct Stack {
 }
 
 impl Stack {
+    #[inline]
     pub fn pop<const N: usize>(&mut self) -> Result<&[U256; N], ExceptionalHalt> {
         // Compile-time check for stack underflow.
-        if N > STACK_LIMIT {
-            return Err(ExceptionalHalt::StackUnderflow);
+        const {
+            assert!(N <= STACK_LIMIT);
         }
 
         // The following operation can never overflow as both `self.offset` and N are within
         // STACK_LIMIT (1024).
-        #[expect(clippy::arithmetic_side_effects)]
-        let next_offset = self.offset + N;
+        let next_offset = self.offset.wrapping_add(N);
 
         // The index cannot fail because `self.offset` is known to be valid. The `first_chunk()`
         // method will ensure that `next_offset` is within `STACK_LIMIT`, so there's no need to
         // check it again.
-        #[expect(clippy::indexing_slicing)]
-        let values = self.values[self.offset..]
-            .first_chunk::<N>()
-            .ok_or(ExceptionalHalt::StackUnderflow)?;
+        #[expect(unsafe_code)]
+        let values = unsafe {
+            self.values
+                .get_unchecked(self.offset..)
+                .first_chunk::<N>()
+                .ok_or(ExceptionalHalt::StackUnderflow)?
+        };
         self.offset = next_offset;
```

Comment:
I think a comment here mentioning that due to the previous error check `next_offset` is guaranteed to be `< STACK_LIMIT` (due to `first_chunk`).